src/libstd/sys/unix/ext/net.rs

```diff
@@ -0,0 +1,1225 @@
+    {
+        unsafe {
+            let mut addr: libc::sockaddr_un = mem::zeroed();
+            let mut len = mem::size_of::<libc::sockaddr_un>() as libc::socklen_t;
+            try!(cvt(f(&mut addr as *mut _ as *mut _, &mut len)));
+
+            if len == 0 {
+                // When there is a datagram from unnamed unix socket
+                // linux returns zero bytes of address
+                len = sun_path_offset() as libc::socklen_t;  // i.e. zero-length address
+            } else if addr.sun_family != libc::AF_UNIX as libc::sa_family_t {
+                return Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                          "file descriptor did not correspond to a Unix socket"));
+            }
+
+            Ok(SocketAddr {
+                addr: addr,
+                len: len,
+            })
+        }
+    }
+
+    /// Returns true iff the address is unnamed.
+    pub fn is_unnamed(&self) -> bool {
+        if let AddressKind::Unnamed = self.address() {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Returns the contents of this address if it is a `pathname` address.
+    pub fn as_pathname(&self) -> Option<&Path> {
+        if let AddressKind::Pathname(path) = self.address() {
+            Some(path)
+        } else {
+            None
+        }
+    }
+
+    fn address<'a>(&'a self) -> AddressKind<'a> {
+        let len = self.len as usize - sun_path_offset();
+        let path = unsafe { mem::transmute::<&[libc::c_char], &[u8]>(&self.addr.sun_path) };
+
+        // OSX seems to return a len of 16 and a zeroed sun_path for unnamed addresses
+        if len == 0 || (cfg!(not(target_os = "linux")) && self.addr.sun_path[0] == 0) {
+            AddressKind::Unnamed
+        } else if self.addr.sun_path[0] == 0 {
+            AddressKind::Abstract(&path[1..len])
+        } else {
+            AddressKind::Pathname(OsStr::from_bytes(&path[..len - 1]).as_ref())
+        }
+    }
+}
+
+impl fmt::Debug for SocketAddr {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self.address() {
+            AddressKind::Unnamed => write!(fmt, "(unnamed)"),
+            AddressKind::Abstract(name) => write!(fmt, "{} (abstract)", AsciiEscaped(name)),
+            AddressKind::Pathname(path) => write!(fmt, "{:?} (pathname)", path),
+        }
+    }
+}
+
+struct AsciiEscaped<'a>(&'a [u8]);
+
+impl<'a> fmt::Display for AsciiEscaped<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(fmt, "\""));
+        for byte in self.0.iter().cloned().flat_map(ascii::escape_default) {
+            try!(write!(fmt, "{}", byte as char));
+        }
+        write!(fmt, "\"")
+    }
+}
+
+/// A Unix stream socket.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// #![feature(unix_socket)]
+/// use std::os::unix::net::UnixStream;
+/// use std::io::prelude::*;
+///
+/// # fn main() {
```

Comment:
I think we have special logic to recognize `#![feature]` and do the right thing though?